### DISCUSSION TITLE
v3.2.0: Rails controller integration

### DISCUSSION THREAD

**Discussion by okuramasafumi** [okuramasafumi/alba#373] [Feedback]
Hi, here's a new version of Alba, v3.2.0. There are two additions in this release.

Now we have a Rails "controller" integration! `serialize` and `render_serialized_json` methods are available in Rails controllers, making it easier to use Alba with Rails. It's inspired by [Layered Design for Ruby on Rails Applications](https://www.packtpub.com/en-us/product/layered-design-for-ruby-on-rails-applications-9781801813785) book mentioning Alba and its usage. Thank you!
Another feature is `transform_keys!` that enables to change key transformation type after defining a resource/serializer class. It helps when there are so many classes that are almost equivalent. Note that this feature is experimental.

The documentation for the new Rails controller integration is [here](https://github.com/okuramasafumi/alba/blob/main/docs/rails.md) and the documentation for `transform_keys!` can be found [here](https://github.com/okuramasafumi/alba/blob/main/README.md#experimental-modification-api).Enjoy!

<hr /><em>This discussion was created from the release <a href='https://github.com/okuramasafumi/alba/releases/tag/v3.2.0'>v3.2.0: Rails controller integration</a>.</em>

**Comment by amitpatelx** [comment]
@okuramasafumi I get `undefined method 'render_serialized_json' for an instance of ProductsController`

Looking at the `railtie.rb`, it seems only `ActionController::Base` is the only target for the helper methods
 
https://github.com/okuramasafumi/alba/blob/4072c3b6a978fcd0365d6105cab975e1bc89f559/lib/alba/railtie.rb#L9-L17

curious why you opt out supporting these helpers for `ActionController::API`

**Reply by okuramasafumi** [reply]
Thank you for pointing out, I think we should support `AbstractController` here.

**Reply by okuramasafumi** [reply]
@amitpatelx https://rubygems.org/gems/alba/versions/3.3.2 version 3.3.2 fixes this issue.

**Reply by amitpatelx** [reply]
@okuramasafumi thanks to you for the lovely gem and quick fix 🙏

**Comment by wacci6** [comment]
@okuramasafumi 
It would be great if `render_serialized_json` could also serialize metadata!

**Reply by okuramasafumi** [reply]
@wacci6 What do you mean by "serialize metadata"? Could you provide your ideal API/code?

**Reply by wacci6** [reply]
@okuramasafumi 
すみません、英語が苦手なので日本語でも良いでしょうか 🙏 
GPT翻訳になりますが、一応英語も末尾に載せておきます！

ページネーションの情報をmetadataに付け加えたいと思っていまして。
新しく追加された `render_serialized_json` を使って↓のように書けると便利だなという感じです！

```ruby
# 理想
messages = Message.all
render_serialized_json(messages, with: MessageSerializer, meta: { current_page: 1, next_page: 2, prev_page: nil, ... })

# 現状
messages = Message.all
render json: MessageSerializer.new(messages).serialize(meta: { current_page: 1, next_page: 2, prev_page: nil, ... })
```

最近使い始めたばかりなので、もしすでに理想を実現できる方法がございましたら、ドキュメントか説明箇所のリンクを頂けると助かります 🙇 

---

I would like to include pagination information in the metadata.

It would be really convenient if I could write something like the following using the newly added `render_serialized_json` !

```ruby
# Ideal
messages = Message.all
render_serialized_json(messages, with: MessageSerializer, meta: { current_page: 1, next_page: 2, prev_page: nil, ... })

# Current
messages = Message.all
render json: MessageSerializer.new(messages).serialize(meta: { current_page: 1, next_page: 2, prev_page: nil, ... })
```

I've just started using it recently, so if there's already a way to achieve this, I would appreciate it if you could share a link to the documentation or explanation!

**Reply by okuramasafumi** [reply]
I see. Your request sounds reasonable, so I've implemented this feature in [this PR](https://github.com/okuramasafumi/alba/pull/398). Check it out!

**Reply by wacci6** [reply]
@okuramasafumi 

早速のご対応ありがとうございます！！
思っていた通りの書き方が実現できそうです！リリース楽しみにしてます 😄 

---

Thank you for your prompt response!
It looks like I’ll be able to implement it just as I envisioned! I'm looking forward to the release!
